### PR TITLE
 plugins/crates: init

### LINKED ISSUE #1149: [PLUGIN REQUEST] crates.nvim
| Field    | Description |
|----------|-------------|
| Plugin   | `crates.nvim`  | 
| Homepage | `https://github.com/Saecki/crates.nvim`     | 
| Nixpkgs  | `vimPlugins.crates-nvim` | 

## Extra Information

The plugin has been partially added already but only as a completion source for cmp. It is currently lacking the overlays that show the current versions of the crates when editing a Cargo.toml as you can see in the video on the repo page: https://github.com/Saecki/crates.nvim

### PR DESCRIPTION
**Pull request by GaetanLepage**
Add support for [crates.nvim](https://github.com/Saecki/crates.nvim/tree/main).

I am not convinced on whether I should explicitly declare some options or none...

Fixes #1149

### CODE REVIEW

File: plugins/languages/rust/crates.nix [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+with lib;
+  helpers.neovim-plugin.mkNeovimPlugin config {
+    name = "crates";
+    originalName = "crates.nvim";
+    defaultPackage = pkgs.vimPlugins.crates-nvim;
+
+    maintainers = [maintainers.GaetanLepage];
+
+    settingsOptions = {
+      on_attach = helpers.defaultNullOpts.mkLuaFn "function(bufnr) end" ''
```
Comment (traxys) [inline plugins/languages/rust/crates.nix:17]: Shouldn't we default this to the __lspOnAttach function? I think we should expose this in the helpers in some way, to help  people reuse the default function when creating their own on attach

**Review by traxys — COMMENTED** [review]
The wiki seems to put forwards:
- the `lsp` options
- the `nvim-cmp` completion source (it seems we need to do more that add it to `sources for that!)
- the `coq.nvim`  completion  source
- The null-ls source

So I'd suggest only adding those.
https://github.com/Saecki/crates.nvim/wiki/Documentation-unstable

File: plugins/languages/rust/crates.nix [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        on_attach = helpers.defaultNullOpts.mkLuaFn "function(client, bufnr) end" ''
+          `fun(client: lsp.Client, bufnr: integer)`
+
+          Callback to run when the in-process language server attaches to a buffer.
+
+          NOTE: Ignored if `autoload` is disabled.
+        '';
+
+        # on_attach = helpers.defaultNullOpts.mkLuaFn "__lspOnAttach" ''
+        # '';
```
Comment (traxys) [inline plugins/languages/rust/crates.nix:65]: This should not be present (And like mentionned we should do something about setting the lsp default value to __lspOnAttach to benefit from the other options, like format.nvim, or LSP specific mappings that we could introduce to solve  #1219))

**Review by traxys — CHANGES_REQUESTED** [review]

**Comment by GaetanLepage** [comment]
@khaneliman if you happen to want to pick this up, go on :)

**Comment by Eveeifyeve** [comment]
I could look at making changes to this pr as there is two version of crates both cmp and non cmp.

**Comment by khaneliman** [comment]
> I could look at making changes to this pr as there is two version of crates both cmp and non cmp, to help out if needed.

Feel free if you're interested. I can jump in if needed,

**Comment by Eveeifyeve** [comment]
Is there any way of good testing?

**Comment by MattSturgeon** [comment]
> Is there any way of good testing?

You can use `nix flake lock --override-input` in order to temporarily update a nixvim config's flake.lock to use a different input url.

For example:
```bash
# From a nixvim _config_ repo


# Use a local repo
nix flake lock --override-input nixvim ../nixvim-fork

# Use a PR branch
nix flake lock --override-input nixvim github:GaetanLepage/nixvim/crates

# Undo override
git restore flake.lock
```

You can also run nixvim's tests using the `tests` command while in a nixvim devshell:
```bash
# From a nixvim repo


# Enter devshell
nix develop

# Run tests for this plugin (interactive filter)
tests -i
> crates
```

**Comment by Eveeifyeve** [comment]
I will resolve the conflict and everything on this weekend.

**Comment by MattSturgeon** [comment]
I'm going to close this for now. We have [`plugins.crates`](https://nix-community.github.io/nixvim/plugins/crates/index.html?highlight=crates#crates) since #2717, and this PR is so old the conflicts will be a nightmare.

If there's anything of value still in this PR, then new PRs can be created to improve or expand the `plugins.crates` module.

**Pull request closed without merging**